pkg/theme/text.go

```diff
@@ -0,0 +1,395 @@
+package theme
+
+import (
+	"github.com/gookit/color"
+)
+
+var (
+	// DefaultTextColor is the default text color
+	DefaultTextColor = NewTextStyle(color.FgWhite, 0)
+	// DefaultHiTextColor is the default highlighted text color
+	DefaultHiTextColor = NewTextStyle(color.FgLightWhite, 0)
+
+	// SelectedLineBgColor is the background color for the selected line
+	SelectedLineBgColor = NewTextStyle(0, 0)
+
+	// SelectedRangeBgColor is the background color of the selected range of lines
+	SelectedRangeBgColor = NewTextStyle(0, 0)
+
+	OptionsFgColor = NewTextStyle(0, 0)
+
+	DiffTerminalColor = NewTextStyle(color.FgMagenta, 0)
+
+	FgCyan    = NewTextStyle(color.FgCyan, 0)
+	FgRed     = NewTextStyle(color.FgRed, 0)
+	FgGreen   = NewTextStyle(color.FgGreen, 0)
+	FgBlue    = NewTextStyle(color.FgBlue, 0)
+	FgYellow  = NewTextStyle(color.FgYellow, 0)
+	FgMagenta = NewTextStyle(color.FgMagenta, 0)
+)
+
+type TextColorType uint8
+
+const (
+	NoTextColor = TextColorType(iota)
+	BasicTextColor
+	RGBTextColor
+)
+
+type TextStyle interface {
+	Sprint(a ...interface{}) string
+	Sprintf(format string, a ...interface{}) string
+	SetBold(v bool) TextStyle
+	SetReverse(v bool) TextStyle
+	SetUnderline(v bool) TextStyle
+	SetColor(color color.Color, background bool) TextStyle
+	SetRGBColor(r, g, b uint8, background bool) TextStyle
+	Merge(other TextStyle) TextStyle
+	GetOpts() color.Opts
+}
+
+func NewTextStyle(fg color.Color, bg color.Color, opts ...color.Color) TextStyle {
+	res := BasicTextStyle{}
+
+	if fg != 0 {
+		res.hasFg = true
+		res.style = append(res.style, fg)
+	}
+	if bg != 0 {
+		res.hasBg = true
+		res.style = append(res.style, bg)
+	}
+
+	for _, opt := range opts {
+		switch opt {
+		case color.OpBold, color.OpReverse, color.OpUnderscore:
+			res.style = append(res.style, opt)
+		}
+	}
+
+	return res
+}
+
+type BasicTextStyle struct {
+	hasFg bool
+	hasBg bool
+	style color.Style
+}
+
+func (b BasicTextStyle) Sprint(a ...interface{}) string {
+	return b.style.Sprint(a...)
+}
+
+func (b BasicTextStyle) Sprintf(format string, a ...interface{}) string {
+	return b.style.Sprintf(format, a...)
+}
+
+func (b BasicTextStyle) setOpt(opt color.Color, v bool) BasicTextStyle {
+	optsOffset := b.getOpsOffset()
+	opts := b.style[optsOffset:]
+	if v {
+		// Add value
+		for _, listOpt := range opts {
+			if listOpt == opt {
+				return b
+			}
+		}
+		b.style = append(b.style, opt)
+	} else {
+		// Remove value
+		for idx, listOpt := range opts {
+			if listOpt == opt {
+				b.style = append(b.style[:optsOffset+idx], b.style[optsOffset+idx+1:]...)
+				return b
+			}
+		}
+	}
+	return b
+}
+
+func (b BasicTextStyle) SetBold(v bool) TextStyle {
+	return b.setOpt(color.OpBold, v)
+}
+
+func (b BasicTextStyle) SetReverse(v bool) TextStyle {
+	return b.setOpt(color.OpReverse, v)
+}
+
+func (b BasicTextStyle) SetUnderline(v bool) TextStyle {
+	return b.setOpt(color.OpUnderscore, v)
+}
+
+func (b BasicTextStyle) SetColor(v color.Color, background bool) TextStyle {
```

Comment:
this function is quite complex and I had to read it a few times to understand it.
We could simplify this by having our BasicTextStyle struct store fg, bg, and opts fields explicitly, and then only when we call Sprint (or similar functions) we generate a Color slice from those fields. This has the downside of being more expensive at runtime, because we'll be calling Sprint much more than we'll be calling SetColor. But we could have the best of both worlds by introducing a style field which is derived from fg/bg/opts whenever any of those fields changes. For example:

```go
type BasicTextStyle struct {
	// using a pointer because the zero value of color.Color is color.OpReset which I suspect
	// is not the same as saying 'I don't want to set a foreground color'
	fg   *color.Color
	bg   *color.Color
	opts []color.Color

	// we derive this from the fg, bg, and opts fields. We could derive style
	// on the fly but that might be expensive
	style color.Style
}

func (b BasicTextStyle) deriveStyle() {
	style := make([]color.Color, 0, 2+len(b.opts))

	if b.fg != nil {
		style = append(style, *b.fg)
	}

	if b.bg != nil {
		style = append(style, *b.bg)
	}

	style = append(style, b.opts...)

	b.style = style
}

func (b BasicTextStyle) setOpt(opt color.Color, v bool) BasicTextStyle {
	opts := b.opts
	if v {
		// Add value
		for _, listOpt := range opts {
			if listOpt == opt {
				return b
			}
		}
		b.opts = append(b.opts, opt)
	} else {
		// Remove value
		for idx, listOpt := range opts {
			if listOpt == opt {
				b.opts = append(b.opts[:idx], b.opts[idx+1:]...)
				return b
			}
		}
	}

	b.deriveStyle()

	return b
}

func (b BasicTextStyle) SetColor(v color.Color, background bool) TextStyle {
	if background {
		b.bg = &v
	} else {
		b.fg = &v
	}

	b.deriveStyle()

	return b
}
```

What do you think?

Reply:
> using a pointer because the zero value of color.Color is color.OpReset which I suspect

We should then also use pointers in the creation of a new style but the color library static values cannot be used as a pointer directly and thus require some sort of extra code to make it work.  
I would say as the `OpReset` is a value for the `opts` array we should just use null values for the fg and bg colors

```
cannot use color.FgCyan (constant 36 of type color.Color) as *color.Color value in argument to New
```

Reply:
Have changed the BasicTextStyle to somewhat represent your idea here.